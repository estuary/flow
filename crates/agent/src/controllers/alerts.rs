use std::collections::HashMap;

use chrono::{DateTime, Utc};
use models::status::{AlertState, AlertType, Alerts, ControllerAlert};

pub fn set_alert_firing(
    statuses: &mut Alerts,
    alert_type: AlertType,
    now: DateTime<Utc>,
    error: String,
    count: u32,
    spec_type: models::CatalogType,
) {
    // If the current status is for a resolved alert of the same type, remove
    // that entry and re-start with a fresh alert.
    if statuses
        .get(&alert_type)
        .is_some_and(|s| s.state == AlertState::Resolved)
    {
        statuses.remove(&alert_type);
    }

    if let Some(existing) = statuses.get_mut(&alert_type) {
        existing.last_ts = Some(now);
        existing.count = count;
        existing.spec_type = spec_type;
        existing.error = error;
    } else {
        tracing::info!(%alert_type, %spec_type, "alert started firing");
        statuses.insert(
            alert_type,
            ControllerAlert {
                state: AlertState::Firing,
                first_ts: now,
                last_ts: None,
                error,
                count,
                spec_type,
                resolved_at: None,
                extra: HashMap::new(),
            },
        );
    }
}

/// Clears the given alert entirely, removing it from `statuses`.
/// This does not allow for the presence of separate `resolved_arguments`.
/// If separate `resovlved_arguments` are needed, then we'll have to
pub fn resolve_alert(statuses: &mut Alerts, alert_type: AlertType) {
    if let Some(_cleared) = statuses.remove(&alert_type) {
        tracing::info!(%alert_type, "alert resolved");
    }
}
