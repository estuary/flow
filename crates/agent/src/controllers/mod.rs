mod capture;
mod collection;
mod handler;
//mod inferred_schema;
mod materialization;
mod publication_status;
//mod source_capture;
mod test;

#[cfg(test)]
pub mod test_util;

use crate::{controlplane::ControlPlane, publications::PublicationResult};
use anyhow::Context;
use chrono::{DateTime, Utc};
use itertools::Itertools;
use models::Id;
use serde::{Deserialize, Serialize};
use sqlx::types::Uuid;
use std::fmt::Debug;
use tables::AnySpec;

use self::{
    capture::CaptureStatus,
    collection::CollectionStatus,
    materialization::MaterializationStatus,
    publication_status::{PendingPublication, PublicationInfo, PublicationStatus},
    test::TestStatus,
};

pub use handler::ControllerHandler;

/// This version is used to determine if the controller state is compatible with the current
/// code. Any controller state having a lower version than this will need to be run in order
/// to "upgrade" it. Any controller state having a higher version than this _must_ be ignored.
///
/// Increment this version whenever we need to ensure that controllers re-visit all live specs.
pub const CONTROLLER_VERSION: i32 = 1;

/// Represents the state of a specific controller and catalog_name.
#[derive(Clone, Debug, Serialize)]
pub struct ControllerState {
    pub catalog_name: String,
    pub live_spec: Option<AnySpec>,
    pub next_run: Option<DateTime<Utc>>,
    pub updated_at: DateTime<Utc>,
    pub failures: i32,
    pub errror: Option<String>,
    pub last_pub_id: Id,
    pub logs_token: Uuid,
    pub controller_version: i32,
    pub current_status: Status,
}

impl ControllerState {
    pub fn parse_db_row(
        job: &agent_sql::controllers::ControllerJob,
    ) -> anyhow::Result<ControllerState> {
        let status: Status = if job.controller_version == 0 {
            Status::Uninitialized
        } else {
            serde_json::from_str(job.status.get()).context("deserializing controller status")?
        };
        let catalog_type = job.spec_type.into();

        let live_spec = if let Some(live_json) = &job.live_spec {
            let spec = tables::AnySpec::deserialize(catalog_type, &live_json)
                .context("deserializing live spec")?;
            Some(spec)
        } else {
            None
        };
        let controller_state = ControllerState {
            next_run: job.controller_next_run,
            updated_at: job.updated_at,
            live_spec,
            failures: job.failures,
            catalog_name: job.catalog_name.clone(),
            errror: job.error.clone(),
            last_pub_id: job.last_pub_id.into(),
            logs_token: job.logs_token,
            controller_version: job.controller_version,
            current_status: status,
        };
        Ok(controller_state)
    }
}

#[derive(Debug, Serialize, Clone, Copy, PartialEq, Eq)]
pub struct NextRun {
    pub after_seconds: u32,
    pub jitter_percent: u16,
}

impl std::cmp::PartialOrd for NextRun {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl std::cmp::Ord for NextRun {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.after_seconds
            .cmp(&other.after_seconds)
            .then(self.jitter_percent.cmp(&other.jitter_percent))
    }
}

impl NextRun {
    const DEFAULT_JITTER: u16 = 20;

    pub fn after_minutes(minutes: u32) -> NextRun {
        NextRun {
            after_seconds: minutes * 60,
            jitter_percent: NextRun::DEFAULT_JITTER,
        }
    }

    pub fn with_jitter_percent(self, jitter_percent: u16) -> Self {
        NextRun {
            after_seconds: self.after_seconds,
            jitter_percent,
        }
    }

    /// Returns an absolute time at which the next run should become due.
    /// Uses only millisecond precision to ensure that the timestamp can be losslessly
    /// round-tripped through postgres.
    pub fn compute_time(&self) -> DateTime<Utc> {
        use rand::Rng;

        let delta_millis = self.after_seconds as i64 * 1000;
        let jitter_mul = self.jitter_percent as f64 / 100.0;
        let jitter_max = (delta_millis as f64 * jitter_mul) as i64;
        let jitter_add = rand::thread_rng().gen_range(0..jitter_max);
        let dur = chrono::TimeDelta::milliseconds(delta_millis + jitter_add);
        Utc::now() + dur
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum Status {
    Capture(CaptureStatus),
    Collection(CollectionStatus),
    Materialization(MaterializationStatus),
    Test(TestStatus),
    #[serde(other, untagged)]
    Uninitialized,
}

impl Status {
    async fn update<C: ControlPlane>(
        &mut self,
        state: &ControllerState,
        control_plane: &mut C,
    ) -> anyhow::Result<Option<NextRun>> {
        let Some(live_spec) = &state.live_spec else {
            tracing::info!("live spec has been deleted, notifying dependents");
            // The live spec has been deleted. Notify dependent controllers, to make sure that they
            // can respond. The controller job row will be deleted automoatically after we return.
            control_plane
                .notify_dependents(state.catalog_name.clone(), state.last_pub_id)
                .await?;
            return Ok(None);
        };

        let next_run = match live_spec {
            AnySpec::Capture(c) => {
                let capture_status = self.as_capture_mut()?;
                capture_status.update(state, control_plane, c).await?
            }
            AnySpec::Collection(c) => {
                let collection_status = self.as_collection_mut()?;
                collection_status.update(state, control_plane, c).await?
            }
            AnySpec::Materialization(m) => {
                let materialization_status = self.as_materialization_mut()?;
                materialization_status
                    .update(state, control_plane, m)
                    .await?
            }
            AnySpec::Test(t) => {
                let test_status = self.as_test_mut()?;
                test_status.update(state, control_plane, t).await?
            }
        };
        tracing::info!(?next_run, "finished controller update");
        Ok(next_run)
    }

    pub fn is_uninitialized(&self) -> bool {
        matches!(self, Status::Uninitialized)
    }

    fn as_capture_mut(&mut self) -> anyhow::Result<&mut CaptureStatus> {
        if self.is_uninitialized() {
            *self = Status::Capture(Default::default());
        }
        match self {
            Status::Capture(c) => Ok(c),
            _ => anyhow::bail!("expected capture status"),
        }
    }

    fn as_collection_mut(&mut self) -> anyhow::Result<&mut CollectionStatus> {
        if self.is_uninitialized() {
            *self = Status::Collection(Default::default());
        }
        match self {
            Status::Collection(c) => Ok(c),
            _ => anyhow::bail!("expected collection status"),
        }
    }

    fn as_materialization_mut(&mut self) -> anyhow::Result<&mut MaterializationStatus> {
        if self.is_uninitialized() {
            *self = Status::Materialization(Default::default());
        }
        match self {
            Status::Materialization(m) => Ok(m),
            _ => anyhow::bail!("expected materialization status"),
        }
    }

    fn as_test_mut(&mut self) -> anyhow::Result<&mut TestStatus> {
        if self.is_uninitialized() {
            *self = Status::Test(Default::default());
        }
        match self {
            Status::Test(t) => Ok(t),
            _ => anyhow::bail!("expected test status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_capture(&self) -> &CaptureStatus {
        match self {
            Status::Capture(c) => c,
            _ => panic!("expected capture status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_collection(&self) -> &CollectionStatus {
        match self {
            Status::Collection(c) => c,
            _ => panic!("expected collection status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_materialization(&self) -> &MaterializationStatus {
        match self {
            Status::Materialization(m) => m,
            _ => panic!("expected materialization status"),
        }
    }

    #[cfg(test)]
    pub fn unwrap_test(&self) -> &TestStatus {
        match self {
            Status::Test(t) => t,
            _ => panic!("expected test status"),
        }
    }
}

fn reduce_next_run(next_runs: &[Option<NextRun>]) -> Option<NextRun> {
    let mut min: Option<NextRun> = None;
    for next_run in next_runs {
        match (min, *next_run) {
            (Some(l), Some(r)) => min = Some(l.min(r)),
            (None, Some(r)) => min = Some(r),
            (_, None) => { /* nada */ }
        }
    }
    min
}

#[tracing::instrument(skip_all, err)]
pub async fn finish_pending_publication<C: ControlPlane>(
    pub_status: &mut PublicationStatus,
    pending: PendingPublication,
    state: &ControllerState,
    cp: &mut C,
) -> anyhow::Result<PublicationResult> {
    let detail = pending.details.iter().join(", ");
    let result = cp
        .publish(pending.id, Some(detail), state.logs_token, pending.draft)
        .await?;

    pub_status.record_result(PublicationInfo::observed(&result));
    if result.publication_status.is_success() {
        cp.notify_dependents(state.catalog_name.clone(), result.publication_id)
            .await?;
    }

    Ok(result)
}
