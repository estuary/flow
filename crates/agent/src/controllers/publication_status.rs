use crate::publications;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::VecDeque;
use tables::Id;

use super::PublicationResult;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct PublicationStatus {
    pub id: Id,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<publications::JobStatus>,
}

impl PublicationStatus {
    pub fn is_success(&self) -> bool {
        // TODO: should EmptyDraft be considered successful?
        self.result.as_ref().is_some_and(|s| s.is_success())
    }

    pub fn is_incompatible_collection_error(
        &self,
        collection_name: &str,
    ) -> Option<Vec<publications::builds::AffectedConsumer>> {
        match &self.result {
            Some(publications::JobStatus::BuildFailed {
                incompatible_collections,
                ..
            }) => incompatible_collections
                .iter()
                .find(|ic| ic.collection == collection_name)
                .map(|ic| ic.affected_materializations.clone()),
            _ => None,
        }
    }

    pub fn created(id: Id, time: DateTime<Utc>) -> Self {
        PublicationStatus {
            id,
            created: Some(time),
            completed: None,
            result: None,
        }
    }

    pub fn observed(publication: &PublicationResult) -> Self {
        PublicationStatus {
            id: publication.publication_id,
            created: None,
            completed: Some(publication.completed_at),
            result: Some(publication.publication_status.clone()),
        }
    }

    pub fn with_update(mut self, publication: PublicationStatus) -> Self {
        self.update(publication);
        self
    }

    pub fn update(&mut self, publication: PublicationStatus) {
        assert_eq!(
            self.id, publication.id,
            "complete must be invoked for the same publication id"
        );
        self.completed = publication.completed;
        self.result = publication.result.clone();
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PublicationHistory {
    pub pending: Option<PublicationStatus>,
    pub history: VecDeque<PublicationStatus>,
    pub last_observed: Option<PublicationStatus>,
}

impl PublicationHistory {
    const MAX_HISTORY: usize = 3;

    pub fn initial(observed: PublicationStatus) -> Self {
        PublicationHistory {
            pending: None,
            history: VecDeque::new(),
            last_observed: Some(observed),
        }
    }

    pub fn is_pending(&self, publication_id: Id) -> bool {
        self.pending.iter().any(|p| p.id == publication_id)
    }

    /// Updates status to reflect the recently observed publication. Returns `Some` if this publication
    /// was one that the controller itself has created and was waiting on.
    pub fn observe(&mut self, publication: PublicationStatus) -> Option<PublicationStatus> {
        if self
            .pending
            .as_ref()
            .is_some_and(|p| p.id == publication.id)
        {
            let completed = self.pending.take().unwrap().with_update(publication);
            self.history.push_front(completed.clone());
            while self.history.len() > PublicationHistory::MAX_HISTORY {
                self.history.pop_back();
            }
            self.last_observed = Some(completed.clone());
            Some(completed)
        } else {
            self.last_observed = Some(publication.clone());
            None
        }
    }
}
