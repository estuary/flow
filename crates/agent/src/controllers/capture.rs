use super::{publication_status::Dependencies, ControlPlane, ControllerState, NextRun};
use crate::controllers::publication_status::PublicationStatus;
use chrono::{DateTime, Utc};
use itertools::Itertools;
use serde::{Deserialize, Serialize};

/// placeholder capture status
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct CaptureStatus {
    // TODO: auto discovers are not yet implemented as controllers, but they should be soon.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_auto_discover: Option<DateTime<Utc>>,
    #[serde(default)]
    pub publications: PublicationStatus,
}

impl CaptureStatus {
    pub async fn update<C: ControlPlane>(
        &mut self,
        state: &ControllerState,
        control_plane: &mut C,
        _model: &models::CaptureDef,
    ) -> anyhow::Result<Option<NextRun>> {
        let dependencies = Dependencies::resolve(&state.live_spec, control_plane).await?;
        if dependencies.is_publication_required(state) {
            let draft = self
                .publications
                .start_spec_update(
                    dependencies.next_pub_id(control_plane),
                    state,
                    format!("in response to publication of one or more depencencies"),
                )
                .await?;
            if !dependencies.deleted.is_empty() {
                tracing::debug!(deleted_collections = ?dependencies.deleted, "disabling bindings for collections that have been deleted");
                let draft_capture = draft
                    .captures
                    .get_mut_by_key(&models::Capture::new(&state.catalog_name))
                    .expect("draft must contain capture");
                let mut disabled_count = 0;
                for binding in draft_capture.model.as_mut().unwrap().bindings.iter_mut() {
                    if dependencies.deleted.contains(binding.target.as_str()) && !binding.disable {
                        disabled_count += 1;
                        binding.disable = true;
                    }
                }
                let detail = format!(
                    "disabled {disabled_count} binding(s) in response to deleted collections: [{}]",
                    dependencies.deleted.iter().format(", ")
                );
                self.publications
                    .update_pending_draft(detail, control_plane);
            }
        }

        // TODO: implement auto discover here

        if self.publications.has_pending() {
            let _result = self
                .publications
                .finish_pending_publication(state, control_plane)
                .await?;
            // TODO: check result for incompatible collections and increment backfill counters as necessary
        } else {
            self.publications
                .notify_dependents(state, control_plane)
                .await?;
        }

        Ok(self.publications.next_run(state))
    }
}
