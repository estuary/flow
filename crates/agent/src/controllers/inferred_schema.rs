use std::collections::{BTreeMap, BTreeSet};

use crate::controllers::publication_status::{PublicationHistory, PublicationStatus};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tables::{DraftRow, EitherOrBoth, SpecExt};

use super::{
    ControlJob, ControlPlane, ControllerState, ControllerUpdate, NextRun, PublicationResult,
};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct InferredSchemaStatus {
    pub schema_last_updated: DateTime<Utc>,
    pub schema_md5: Option<String>,
    pub publications: PublicationHistory,
}

impl InferredSchemaStatus {
    fn next_run(&self, now: DateTime<Utc>) -> NextRun {
        // The idea here is to check frequently if there isn't an inferred schema at all yet,
        // so we can quickly start materializing some data. But after it works at least once,
        // then we want to use a longer duration in order to coalesce more schema updates into
        // each publication to prevent undue churn.
        // TODO: we might want to account for `last_backfill` times here
        let min_backoff_minutes = if self.schema_md5.is_none() {
            1i64
        } else {
            10i64
        };

        let after_minutes = now
            .signed_duration_since(self.schema_last_updated)
            .num_minutes()
            .max(min_backoff_minutes)
            .min(90);
        NextRun::after_minutes(after_minutes as u32).with_jitter_percent(25)
    }

    fn is_publication_pending(&self) -> bool {
        self.publications.pending.is_some()
    }

    /// Updates the status of all consumers of the collection to reflect that they're using the latest inferred schema.
    /// This works because if the collection itself is published, then all consumers of it must also be included in the
    /// publication.
    fn on_successful_collection_publication(
        &self,
        _collection_name: &models::Collection,
        schema_md5: Option<String>,
        publication: &PublicationResult,
    ) -> Self {
        let mut new_status = self.clone();
        new_status.schema_md5 = schema_md5.clone();
        new_status.schema_last_updated = publication.completed_at;

        let pub_status = PublicationStatus::observed(publication);
        let _completed = new_status.publications.observe(pub_status.clone());

        new_status
    }

    fn on_failed_publication(
        &self,
        collection_name: &models::Collection,
        publication: &PublicationResult,
    ) -> Self {
        let mut new_status = self.clone();
        let schema_md5 = get_inferred_schema_md5(collection_name, publication);
        // We can still update the last known schema md5, if known.
        if schema_md5.is_some() && schema_md5 != new_status.schema_md5 {
            new_status.schema_md5 = schema_md5.clone();
            new_status.schema_last_updated = publication.completed_at;
        }

        let _ = new_status
            .publications
            .observe(PublicationStatus::observed(publication))
            .expect("publication must be pending");

        new_status
    }
}

fn get_inferred_schema_md5(
    collection: &models::Collection,
    publication: &PublicationResult,
) -> Option<String> {
    publication
        .live
        .inferred_schemas
        .get_by_key(collection)
        .map(|s| s.md5.clone())
}

pub struct InferredSchemaController;

#[async_trait::async_trait]
impl ControlJob for InferredSchemaController {
    type Status = InferredSchemaStatus;

    fn controller_name(&self) -> String {
        String::from("InferredSchemaPublisher")
    }

    fn observe_publication(
        &self,
        current_states: BTreeMap<String, ControllerState<Self::Status>>,
        publication: &PublicationResult,
    ) -> anyhow::Result<BTreeMap<String, ControllerUpdate<Self::Status>>> {
        if publication.publication_status.is_success() {
            Ok(on_successful_publication(&current_states, publication))
        } else {
            Ok(on_failed_publication(&current_states, publication))
        }
    }

    async fn update(
        &self,
        catalog_name: String,
        current_state: ControllerState<Self::Status>,
        control_plane: &mut dyn ControlPlane,
    ) -> anyhow::Result<ControllerUpdate<Self::Status>> {
        if current_state.status.is_publication_pending() {
            // We're still waiting on a publication to complete, so just keep waiting.
            // Note that we don't set next_run or update the status.
            return Ok(current_state.to_update());
        }
        let collection_name = models::Collection::new(catalog_name);

        let maybe_schema_md5 = dbg!(control_plane
            .get_inferred_schema(collection_name.clone())
            .await?
            .map(|s| s.md5));

        if current_state.status.schema_md5 != maybe_schema_md5 {
            let mut new_status = current_state.status.clone();
            create_publication(collection_name, &mut new_status, control_plane).await?;

            Ok(current_state.to_update().with_status(new_status))
        } else {
            eprintln!("wtf everything is up to date");
            // Everything seems up to date, so just schedule the next check-up
            let next_run = current_state.status.next_run(control_plane.current_time());
            Ok(current_state.to_update().with_next_run(next_run))
        }
    }
}

async fn create_publication(
    collection_name: models::Collection,
    new_status: &mut InferredSchemaStatus,
    control_plane: &mut dyn ControlPlane,
) -> anyhow::Result<()> {
    let mut names = BTreeSet::new();
    names.insert(collection_name.clone());
    let collections = control_plane.get_collections(names).await?;
    if collections.is_empty() {
        anyhow::bail!("collection '{collection_name}' not found");
    };

    let catalog = tables::LiveCatalog {
        collections,
        ..Default::default()
    };
    let pub_id = control_plane
        .create_publication(tables::DraftCatalog::from(catalog))
        .await?;
    new_status.publications.pending = Some(PublicationStatus::created(
        pub_id,
        control_plane.current_time(),
    ));
    Ok(())
}

fn on_successful_publication(
    current_states: &BTreeMap<String, ControllerState<InferredSchemaStatus>>,
    publication: &PublicationResult,
) -> BTreeMap<String, ControllerUpdate<InferredSchemaStatus>> {
    let mut updates = BTreeMap::new();

    let drafted_collections = publication
        .draft
        .collections
        .iter()
        .filter(|c| c.spec.as_ref().map(|s| s.is_enabled()).unwrap_or(false))
        .map(|c| (c.catalog_name(), c.spec().unwrap()));

    fn left_outer<L, R>(eob: EitherOrBoth<L, R>) -> Option<(L, Option<R>)> {
        match eob {
            EitherOrBoth::Both(l, r) => Some((l, Some(r))),
            EitherOrBoth::Left(l) => Some((l, None)),
            EitherOrBoth::Right(_) => None,
        }
    }

    fn right_outer<L, R>(eob: EitherOrBoth<L, R>) -> Option<(Option<L>, R)> {
        match eob {
            EitherOrBoth::Both(l, r) => Some((Some(l), r)),
            EitherOrBoth::Left(l) => None,
            EitherOrBoth::Right(r) => Some((None, r)),
        }
    }

    for (maybe_inferred_schema, (collection_name, draft_spec)) in publication
        .live
        .inferred_schemas
        .outer_join(drafted_collections, right_outer)
    {
        let current_state = current_states.get(collection_name.as_str());
        let desired_active = uses_inferred_schema(draft_spec);
        let schema_md5 = maybe_inferred_schema.as_ref().map(|s| s.md5.clone());

        let maybe_update = match (current_state, desired_active) {
            (Some(state), true) => {
                let next_status = state.status.on_successful_collection_publication(
                    collection_name,
                    schema_md5,
                    publication,
                );
                let next_run = next_status.next_run(publication.completed_at);
                Some(
                    state
                        .to_update()
                        .with_status(next_status)
                        .with_next_run(next_run),
                )
            }
            // State is currently active, but the collection no longer uses schema inference, so we should deactivate.
            (Some(state), false) if state.active => Some(state.to_update().with_active(false)),
            (Some(_), false) => None, // inactive state remains inactive
            // Need to initialize a new state for this collection
            (None, true) => {
                let next_status = InferredSchemaStatus {
                    schema_md5,
                    schema_last_updated: publication.completed_at,
                    publications: PublicationHistory::initial(PublicationStatus::observed(
                        publication,
                    )),
                };
                let next_run = next_status.next_run(publication.completed_at);
                Some(ControllerUpdate::new(next_status).with_next_run(next_run))
            }
            (None, false) => None,
        };

        if let Some(update) = maybe_update {
            updates.insert(collection_name.to_string(), update);
        }
    }

    updates
}

fn on_failed_publication(
    current_states: &BTreeMap<String, ControllerState<InferredSchemaStatus>>,
    publication: &PublicationResult,
) -> BTreeMap<String, ControllerUpdate<InferredSchemaStatus>> {
    let mut updates = BTreeMap::new();

    for (collection_name, state) in current_states.iter().filter(|(_, v)| v.active) {
        if state
            .status
            .publications
            .is_pending(publication.publication_id)
        {
            let collection = models::Collection::new(collection_name);
            let next_status = state.status.on_failed_publication(&collection, publication);
            let next_run = next_status.next_run(publication.completed_at);
            let update = state
                .to_update()
                .with_status(next_status)
                .with_next_run(next_run);
            updates.insert(collection_name.clone(), update);
        }
    }

    updates
}

fn uses_inferred_schema(collection: &models::CollectionDef) -> bool {
    collection
        .read_schema
        .as_ref()
        .map(models::Schema::references_inferred_schema)
        .unwrap_or(false)
}

#[cfg(test)]
mod test {
    use crate::controllers::test_util::{self, Harness, TestPublication};
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn happy_path_test() {
        let mut harness = Harness::new(InferredSchemaController);

        let initial_draft = test_util::draft_of(json!({
            "captures": {
                "a/cap": mock_capture_spec(&["acmeCo/a"]),
            },
            "collections": {
                "a/a": mock_collection_spec(true),
                "a/d": mock_derivation_spec(true, &["a/a"]),
                // f should be ignored because it doesn't use schema inference
                "a/f": mock_collection_spec(false),
            },
            "materializations": {
                "a/m1": mock_materialization_spec(&["a/a", "a/d"]),
                "a/m2": mock_materialization_spec(&["a/a"]),
                "a/m3": mock_materialization_spec(&["a/d"]),
            }
        }));

        let (info, updates) = harness.observe_publication(TestPublication::of_draft(initial_draft));
        // Expect to see updates for both "a/a" and "a/d" that initialize new controller states
        // for those collections.
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("initial-updates", &updates);
        });

        // The inferred schema of a/a gets updated.
        harness.update_inferred_schema("a/a", 1);

        // Technically, either a/a or a/d could be run next in the real world, since they both have
        // identical next_run values. Time is deterministic in the test environment, though, and the
        // harness always selects controllers in lexicographical order.
        let (info, update, mut publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/a");
        // Expect to see that a publication of a/a was created in response to the schema being updated
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("update-after-a-schema-updated-1", (update, &publications));
        });

        // Now the controller observes the successful completion of that publication
        let publication = publications.pop().unwrap();
        let (info, updates) = harness.observe_publication(publication);

        // Expect to see the successful publication in the history of a/a
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("observe-pub-a-1-completed", &updates);
        });

        // Now the inferred schema for a/d gets updated
        harness.update_inferred_schema("a/d", 1);

        let (info, update, mut publications) = harness.next_run_update();

        assert_eq!(info.catalog_name, "a/d");
        // Expect to see that a publication of a/d was created in response to the schema being updated
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("update-after-d-schema-updated-1", (update, &publications));
        });

        let (info, updates) = harness.observe_publication(publications.pop().unwrap());

        // Expect to see the successful publication in the history of a/d
        insta::with_settings!({ info => &info }, {
            insta::assert_json_snapshot!("observe-pub-d-1-completed", &updates);
        });

        // Assert that the next update runs don't publish anthing, since the inferred schemas haven't been updated
        let (info, update, publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/a");
        assert!(publications.is_empty());
        assert!(update.next_run.is_some());

        let (info, update, publications) = harness.next_run_update();
        assert_eq!(info.catalog_name, "a/d");
        assert!(publications.is_empty());
        assert!(update.next_run.is_some());
    }

    fn mock_derivation_spec(use_inferred: bool, sources: &[&str]) -> Value {
        let mut collection = mock_collection_spec(use_inferred);

        let transforms = sources
            .iter()
            .map(|name| {
                json!({
                    "name": name, "source": name, "lambda": "shelect shtar"
                })
            })
            .collect::<Vec<_>>();
        let derive = json!({
            "using": {
                "sqlite": {}
            },
            "transforms": transforms,
        });
        collection
            .as_object_mut()
            .unwrap()
            .insert("derive".to_owned(), derive);
        collection
    }

    fn mock_collection_spec(uses_inferred: bool) -> Value {
        let read_schema = if uses_inferred {
            json!({
                "allOf": [
                    { "$ref": "flow://inferred-schema" },
                    { "$ref": "flow://write-schema" },
                ]
            })
        } else {
            json!({"type": "object"})
        };
        json!({
            "key": ["/id"],
            "writeSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"}
                }
            },
            "readSchema": read_schema,
        })
    }

    fn mock_capture_spec(bound_collections: &[&str]) -> Value {
        let bindings = bound_collections
            .into_iter()
            .map(|collection| {
                json!({
                    "resource": { "thing": collection },
                    "target": collection,
                })
            })
            .collect();
        mock_connector_task(bindings)
    }

    fn mock_materialization_spec(bound_collections: &[&str]) -> Value {
        let bindings = bound_collections
            .into_iter()
            .map(|collection| {
                json!({
                    "resource": { "thing": collection },
                    "source": collection,
                })
            })
            .collect();
        mock_connector_task(bindings)
    }

    fn mock_connector_task(bindings: Vec<Value>) -> Value {
        serde_json::json!({
            "endpoint": {
                "connector": {
                    "image": "foo:test",
                    "config": {},
                }
            },
            "bindings": bindings,
        })
    }
}
