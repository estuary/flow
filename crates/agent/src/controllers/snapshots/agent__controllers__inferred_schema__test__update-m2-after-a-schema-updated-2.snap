---
source: crates/agent/src/controllers/inferred_schema.rs
expression: "(update, &publications)"
info:
  catalog_name: a/m2
  prev_state:
    active: true
    next_run: "2024-01-01T08:42:00Z"
    updated_at: "2024-01-01T08:22:00Z"
    status:
      collections:
        a/a:
          inferred_schema_md5: 9a715d81eee48cba22c4469c1417d697
          md5_last_updated: "2024-01-01T08:02:00Z"
      publications:
        pending: ~
        history:
          - id: "0300000000000000"
            created: "2024-01-01T08:02:00Z"
            completed: "2024-01-01T08:02:00Z"
            result:
              type: success
        last_observed:
          id: "0300000000000000"
          created: "2024-01-01T08:02:00Z"
          completed: "2024-01-01T08:02:00Z"
          result:
            type: success
    failures: 0
---
[
  {
    "active": true,
    "next_run": null,
    "status": {
      "collections": {
        "a/a": {
          "inferred_schema_md5": "9a715d81eee48cba22c4469c1417d697",
          "md5_last_updated": "2024-01-01T08:02:00Z"
        }
      },
      "publications": {
        "pending": {
          "id": "0600000000000000",
          "created": "2024-01-01T08:42:00Z"
        },
        "history": [
          {
            "id": "0300000000000000",
            "created": "2024-01-01T08:02:00Z",
            "completed": "2024-01-01T08:02:00Z",
            "result": {
              "type": "success"
            }
          }
        ],
        "last_observed": {
          "id": "0300000000000000",
          "created": "2024-01-01T08:02:00Z",
          "completed": "2024-01-01T08:02:00Z",
          "result": {
            "type": "success"
          }
        }
      }
    }
  },
  [
    {
      "draft": {
        "materializations": {
          "a/m2": {
            "endpoint": {
              "connector": {
                "image": "foo:test",
                "config": {
                  "$serde_json::private::RawValue": "{}"
                }
              }
            },
            "bindings": [
              {
                "resource": {
                  "$serde_json::private::RawValue": "{\"thing\":\"a/a\"}"
                },
                "source": "a/a",
                "fields": {
                  "recommended": true
                }
              }
            ]
          }
        }
      },
      "publication_meta": [
        "0600000000000000",
        "2024-01-01T08:42:00Z"
      ]
    }
  ]
]
